## 1. Logistic Regression


Analize data bankloan.csv

1. Bulid a logistic regression Model
    - Target: Default
    - Fetures: Employ, Debtinc, Creddebt, Otherdebt
1. Interpret The Result
1. Validate the model using accuracy in 20% testing data

In [1]:
import pandas as pd
import numpy as np
import statsmodels.api as sm

In [2]:
bankloan = pd.read_csv('bankloan.csv')
bankloan

age  ed  employ  address  income  debtinc   creddebt   othdebt  default
0     41   3      17       12     176      9.3  11.359392  5.008608        1
1     27   1      10        6      31     17.3   1.362202  4.000798        0
2     40   1      15       14      55      5.5   0.856075  2.168925        0
3     41   1      15       14     120      2.9   2.658720  0.821280        0
4     24   2       2        0      28     17.3   1.787436  3.056564        1
..   ...  ..     ...      ...     ...      ...        ...       ...      ...
695   36   2       6       15      27      4.6   0.262062  0.979938        1
696   29   2       6        4      21     11.5   0.369495  2.045505        0
697   33   1      15        3      32      7.6   0.491264  1.940736        0
698   45   1      19       22      77      8.4   2.302608  4.165392        0
699   37   1      12       14      44     14.7   2.994684  3.473316        0

[700 rows x 9 columns]

Age     : Usia \
Ed      : Jenjang pendidikan \
Employ  : Lama kerja \
Address : Lama tinggal di alamat sekarang \
Income  : Pemasukan per bulan \
Debtinc : Persentase hutang dalam income \
Creddebt: Nominal hutang saat ini (satuan 1000 dollar) \
Othdebt : Nominal hutang dari sumber lain



In [3]:
fitur = ['employ','debtinc','creddebt','othdebt']
target = 'default'

x = bankloan[fitur]
y = bankloan[target]


In [4]:
x.describe()

employ     debtinc    creddebt     othdebt
count  700.000000  700.000000  700.000000  700.000000
mean     8.388571   10.260571    1.553553    3.058209
std      6.658039    6.827234    2.117197    3.287555
min      0.000000    0.400000    0.011696    0.045584
25%      3.000000    5.000000    0.369059    1.044178
50%      7.000000    8.600000    0.854869    1.987567
75%     12.000000   14.125000    1.901955    3.923065
max     31.000000   41.300000   20.561310   27.033600

#### Pemodelan

In [5]:
model = sm.Logit(y,sm.add_constant(x)) #Pendefinisian model

result = model.fit()

print(result.summary())

Optimization terminated successfully.
         Current function value: 0.411165
         Iterations 7
                           Logit Regression Results                           
Dep. Variable:                default   No. Observations:                  700
Model:                          Logit   Df Residuals:                      695
Method:                           MLE   Df Model:                            4
Date:                Thu, 02 Feb 2023   Pseudo R-squ.:                  0.2844
Time:                        13:21:40   Log-Likelihood:                -287.82
converged:                       True   LL-Null:                       -402.18
Covariance Type:            nonrobust   LLR p-value:                 2.473e-48
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -1.2302      0.236     -5.210      0.000      -1.693      -0.767
employ        -0.2436      0.

***
Output yang perlu dibaca:
1. LLR P-Value : 0.00...2473
2. P-value:
    1. const    : 0.000
    2. employ   : 0.000
    3. debtinc  : 0.000
    4. creddebt : 0.000
    5. othdevt  : 0.0940 (karna ini 0,094, maka kita tidak perlu interpretasikan coefnya)
    
3. coef:
    1. employ   : -0.2436
    2. debtinc  :  0.0885
    3. creddebt :  0.5041 

***


Yang dimodelkan adalah resiko default (nilai default 1), resiko seseorang akanmenjadi buruk pembayarannya.

Interpretasi:
1. LLR P-Value : 0.00...2473 < 0.05 (tolak H0) \
    Artinya, Minimal salah satu variabel berpengaruh secara signifikan terhadap resiko default

2. P-value:
    1. **const**    : 0.000 < 0.05 (tolak H0) \
        Artinya model kita memerlukan Intersep.
        
    2. **employ**   : 0.000 < 0.05 (tolak H0) \
        Artinya, Lama kerja berpengaruh secara negatif terhadap resiko default. Semakin lama kerja, semakin rendah defaultnya (employ 0-31)
    3. **debtinc**  : 0.000 < 0.05 (tolak H0) \
        Artinya, Persentase hutang berpengaruh secara positif terhadap resio default. Semakin besar persentasi hutang, semakin tinggi resiko default sesorang, berlaku untuk debtinc 0.4 - 41.3.
    4. **creddebt** : 0.000 < 0.05 (tolak H0) \
        Artinya, Nominal hutang berpengaruh secara positif terhadap resio default. Semakin besar nominal hutang, semakin tinggi resiko default sesorang, berlaku untuk creddebt 0.011 - 20.56.
    5. **othdevt**  : 0.0940 > 0.05 (terima H0) \
        Artinya tidak cukup bukti bahwa nominal hutang dari sumber lain berpengaruh terhadap resiko default.
 <br>

3. coef:
    1. employ   : -0.2436
    2. debtinc  :  0.0885
    3. creddebt :  0.5041 


In [6]:
# Untuk Coef dari logistic regression, kita tidak bisa langsung interpretasikan nilai coefnya. 
# Nilai ini harus kita rubah dulu ke Odd Ratio (OR)

#coef -> OR exp(beta(c-a))

# Pilih angka bebas di dalam rangenya
c = 20
a = 15

OR_employ = np.exp(0.2436*(c-a))
print(OR_employ)

3.380420128015566


In [7]:
c = 20
a = 15

OR_debtinc = np.exp(0.0885*(c-a))
print(OR_debtinc)

1.5565938428137092


In [8]:
c = 20
a = 15

OR_creddebt = np.exp(0.5041*(c-a))
print(OR_creddebt)

12.434812515742879


Ketika lama kerja meningkat sebesar 5 tahun (*dari selisih c - a*), kecenderungan seseorang akan menjadi default menurun sebesar 3.38 kali lipat.

3. coef:
    1. employ   : -0.2436 \
        Ketika lama kerja meningkat sebesar 5 tahun (*dari selisih c - a*), kecenderungan seseorang akan menjadi default menurun sebesar 3.38 kali lipat. (dengan catatan nilai variabel lain tidak berubah)
        
    2. debtinc  :  0.0885 \
        Ketika persentase hutang meningkat sebesar 5%, kecenderungan seseorang akan menjadi default meningkat sebesar 1.529 kali lipat.(dengan catatan nilai variabel lain tidak berubah)

    3. creddebt :  0.5041 \
        Ketika nominal hutang meningkat sebesar 5000 dollar, kecenderungan seseorang akan menjadi default meningkat sebesar 12.43 kali lipat.(dengan catatan nilai variabel lain tidak berubah)

Ada 1 lagi yang perlu di cek dari logistic regressi yaitu Multikollinieritasnya.

Harapannya tidak ada multikol, agar interpretasi di atas bisa berlaku.

In [9]:
#Multicol
from statsmodels.stats.outliers_influence import variance_inflation_factor

def calc_vif(x):
    vif = pd.DataFrame()
    vif['variabel'] = x.columns
    vif['vif'] = [variance_inflation_factor(x.values, i) for i in range(x.shape[1])]
    return vif

calc_vif(x)

variabel       vif
0    employ  2.222753
1   debtinc  3.045977
2  creddebt  2.816577
3   othdebt  4.116876

tidak ada masalah multicollinieritas. jadi iterpretasi bisa digunakan.


VALIDASI MODEL

In [10]:
## Validasi model pada 20% data testing

from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score



In [11]:
# MEMBAGI DATANYA
x_train, x_test, y_train, y_test = train_test_split(
    x,
    y,
    test_size = 0.2, # 20% data testing
    random_state = 2020
)



Training model pada data training saja

In [12]:
sm_logit_train = sm.Logit(y_train,sm.add_constant(x_train))
result_train = sm_logit_train.fit()

Optimization terminated successfully.
         Current function value: 0.408607
         Iterations 7


In [13]:
y_prob = result_train.predict(sm.add_constant(x_test)) #resiko/peluang defaultnya belum bad /goodnya

y_class = np.where(y_prob > 0.5, 1, 0) #treshold 0,5, kelas default atau non default (Maksudnya, yang peluangnya besar atau lebih dari 0,5 akan menjadi 1, sisanya 0)

print('Akurasinya adalah sebesar', accuracy_score(y_test,y_class)*100, '%')

Akurasinya adalah sebesar 82.14285714285714 %


Artinya, jika ada 100 nasabah, sekitar 82 nasabah akan terprediksi (default / non-defaultnya) dengan benar.